File: 39-combination-sum/combination-sum.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum = function(candidates, target) {
    const ans = [];

    function backtrack(idx, tar, combine) {
        // Base case: target reached
        if (tar === 0) {
            ans.push([...combine]);  // push copy
            return;
        }

        // If we reach end or target becomes negative
        if (idx >= candidates.length || tar < 0) {
            return;
        }

        // Include current number
        combine.push(candidates[idx]);
        backtrack(idx, tar - candidates[idx], combine);  // stay at idx (can reuse same number)
        combine.pop();

        // Exclude current number
        backtrack(idx + 1, tar, combine);
    }

    backtrack(0, target, []);
    return ans;
};

File: 39-combination-sum/combination-sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        ans = []

        def backtrack(comb, tar, idx):
            if tar == 0:
                ans.append(comb[:])
                return
            
            if(idx >= len(candidates) or tar < 0):
                return

            comb.append(candidates[idx])
            backtrack(comb, tar - candidates[idx], idx)
            comb.pop()
            backtrack(comb, tar, idx + 1)

        backtrack([], target, 0)
        return ans